segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/OperationProcessor.java

```diff
@@ -392,66 +441,210 @@ private boolean isFatalException(Throwable ex) {
          * @param operation The operation to append.
          */
         void addPending(CompletableOperation operation) {
-            this.pendingOperations.add(operation);
+            synchronized (stateLock) {
+                this.pendingOperations.add(operation);
+            }
+
+            autoCompleteIfNeeded();
         }
 
         /**
-         * Gets a value indicating whether there exist any pending operations in this state.
+         * Creates a checkpoint keyed in with the given CommitArgs. This indicates that the DataFrame represented
+         * by the given args has been sealed and is about to be committed. This checkpoint will mark a point in the
+         * OperationMetadataUpdater that corresponds to the end of that DataFrame.
+         *
+         * @param commitArgs The CommitArgs to create a checkpoint for.
          */
-        boolean hasPending() {
-            return !this.pendingOperations.isEmpty();
+        void checkpoint(DataFrameBuilder.CommitArgs commitArgs) {
```

Comment:
This really needs a different name. It's not clear that this is not refering to a metadata checkpoint.